### PR TITLE
Add SshdContainerTest

### PR DESCRIPTION
**Pull request by olivergondza**

### CODE REVIEW

**Comment by olivergondza** [comment]
Here comes the regressions discovered https://ci.jenkins.io/blue/organizations/jenkins/Tools%2Fdocker-fixtures/detail/PR-13/2/pipeline:

```
The following packages have unmet dependencies:

 openssh-server : Depends: openssh-client (= 1:7.5p1-10) but 1:7.5p1-10ubuntu0.1 is to be installed

                  Recommends: libpam-systemd but it is not going to be installed

                  Recommends: ncurses-term but it is not going to be installed

                  Recommends: xauth but it is not going to be installed

                  Recommends: ssh-import-id but it is not going to be installed

E: Unable to correct problems, you have held broken packages.

The command '/bin/sh -c apt-get update -y &&     apt-get install -y         openssh-server=1:7.5p1-10' returned a non-zero code: 100

--->%---

[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 18.792 s <<< FAILURE! - in org.jenkinsci.test.acceptance.docker.JavaContainerTest

[ERROR] smokes(org.jenkinsci.test.acceptance.docker.JavaContainerTest)  Time elapsed: 12.392 s  <<< ERROR!

java.lang.Error: Failed to build image (100): ff2527313f5c

	at org.jenkinsci.test.acceptance.docker.JavaContainerTest.smokes(JavaContainerTest.java:75)

```

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java [hunk 1]
```diff
@@ -53,11 +55,27 @@ public DockerRule(Class<T> type) {
         return this;
     }
 
+    /**
+     * The rule cause the test to be skipped when docker is not available on the host. Setting this to true will cause
+     * the tests to fail.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java]: I do not think this is a good idea. The test should not know about the environment it runs in. Rather, a CI job (or whatever) should be configured to highlight skipped tests.

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile [hunk 2]
```diff
@@ -7,11 +7,11 @@ FROM ubuntu:artful-20171019
 # install SSHD
 RUN apt-get update -y && \
     apt-get install -y \
-        openssh-server=1:7.5p1-10
+        openssh-server=1:7.5p1-10ubuntu0.1
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile]: Seems unrelated?

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile [hunk 3]
```diff
@@ -24,4 +24,6 @@ RUN echo "KexAlgorithms curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha
 RUN echo "MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-ripemd160-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256,hmac-ripemd160,umac-128@openssh.com,hmac-md5,hmac-sha1,hmac-sha1-96,hmac-md5-96" >> /etc/ssh/sshd_config
 
 # run SSHD in the foreground with error messages to stderr
-ENTRYPOINT ["/usr/sbin/sshd", "-D", "-e"]
+ENTRYPOINT ["/usr/sbin/sshd", "-D", "-e", "-ddd"]
+
+RUN getent passwd
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile]: What is this for?

File: src/test/java/org/jenkinsci/test/acceptance/docker/fixtures/JavaContainerTest.java [hunk 4]
```diff
@@ -72,9 +74,10 @@ public static void cleanOldImages() throws Exception {
 
     @Test
     public void smokes() throws Exception {
-        JavaContainer c = rule.get();
-        assertThat(c.popen(new CommandBuilder("java", "-version")).verifyOrDieWith("could not launch Java"), containsString("openjdk version \"1.8.0_"));
+        JavaContainer c = container.get();
+        // System.out.println(c.popen(new CommandBuilder("update-alternatives --display java")).verifyOrDieWith("FAILED"));
+        String version = c.popen(new CommandBuilder("java", "-version")).verifyOrDieWith("could not launch Java");
+        assertThat(version, containsString("openjdk version \"1.8.0_"));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/test/acceptance/docker/fixtures/JavaContainerTest.java]: ?

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java (hunk 1 above)
Comment (olivergondza) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java]: I understand that users to this tooling might feel that way but for testing this repo, I thing this is essential. When I started digging in this, both containers was unbuildable without anyone noticing...

---

I do not insist doing it this way. Though, what is wrong with test author declaring the docker based tests are so essential for the project that when they can not run they should fail the build?

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile (hunk 2 above)
Comment (olivergondza) [reply src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile]: That is the reason the container does no longer build so I am fixing things as I find them broken.

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile (hunk 3 above)
Comment (olivergondza) [reply src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile]: Chasing "No user exists for uid 1000" from ci.jenkins.io  I can not reproduce locally.

File: src/test/java/org/jenkinsci/test/acceptance/docker/fixtures/JavaContainerTest.java (hunk 4 above)
Comment (olivergondza) [reply src/test/java/org/jenkinsci/test/acceptance/docker/fixtures/JavaContainerTest.java]: Ditto

**Comment by olivergondza** [comment]
Sorry for the confusion, adding WIP prefix until this is running and polished.

File: src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile (hunk 2 above)
Comment (jglick) [reply src/main/resources/org/jenkinsci/test/acceptance/docker/fixtures/SshdContainer/Dockerfile]: This part is covered in #14.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java (hunk 1 above)
Comment (jglick) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java]: > both containers was unbuildable without anyone noticing

Trying to address in #15.

Comment (jglick) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java]: > what is wrong with test author declaring the docker based tests are so essential for the project that when they can not run they should fail the build?

Just feels like something that should be defined either in the CI job, or in the build script (e.g., Surefire configuration), or using something like JUnit categories. The same requirement could apply to _any_ kind of test which might be skipped under some conditions (missing Internet connection, ad nauseam).

File: Jenkinsfile [hunk 5]
```diff
@@ -4,15 +4,17 @@ pipeline {
         timeout(time: 1, unit: 'HOURS')
     }
     agent {
-        docker {
-            image 'maven:3.5.0-jdk-8'
+        dockerfile {
```
Comment (jglick) [inline Jenkinsfile]: As the main author of `docker-workflow`, I do not trust it at all, which is why in #15 I am ceasing to use it.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java (hunk 1 above)
Comment (olivergondza) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerRule.java]: Would a java property be any better? I am happy with the PR as it is now ...

**Comment by olivergondza** [comment]
@jglick, removing the controversial part to get the test in.
